### PR TITLE
Update Canceled task flow

### PR DESCRIPTION
**Pull request by thienlnam**
### Fixed Issues

$ https://github.com/Expensify/App/issues/19504
PROPOSAL: 

### Tests

1. Create a task
2. Open the task
3. Cancel the task
4. Verify that the task report shows [Deleted Task] and that you can still comment and interact with the deleted task like a regular thread
5. Verify that in the parent report, the task shows [Deleted Task]

### Offline tests

Same as online

### QA Steps

Same as tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/30609178/840a9731-bd77-40fb-ba86-117f372344e7

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/30609178/9fa2a745-2cd2-458b-b2c7-7463221eb051

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/30609178/4a9ad242-8b13-42ea-8706-476526215d0d

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/30609178/7a5bab6b-3721-4412-bbdb-bd00a0fe63c3

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/30609178/26bcbc3d-d639-4f4d-bd04-1d0e39140aaa

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

File: src/libs/ReportActionsUtils.js [hunk 1]
```diff
@@ -576,7 +576,7 @@ function isMessageDeleted(reportAction) {
     return lodashGet(reportAction, ['message', 0, 'isDeletedParentAction'], false);
 }
 
-/**
+/*
```
Comment (mollfpr) [inline src/libs/ReportActionsUtils.js:579]: NAB: This might be deleted by mistake.

**Review by mollfpr — COMMENTED** [review]
I'm testing the PR now.

**Comment by mollfpr** [comment]
This PR has a different result from the attached design.

|Design|Current|
|---|---|
|Screenshot 2023-09-11 at 21 37 30|Screenshot 2023-09-11 at 21 37 37|


1. There's no line below the parent action/deleted task
2. We are still showing the message of deleted task and the task title

Should we resolve the above note in this PR, or will it be fixed in another issue?

File: src/libs/ReportUtils.js [hunk 2]
```diff
@@ -193,23 +193,37 @@ function isTaskReport(report) {
...
+ * There's another situation where you don't have access to the parentReportAction (because it was created in a chat you don't have access to)
+ * In this case, we have added the key to the report itself
  *
  * @param {Object} report
+ * @param {Object} parentReportAction
  * @returns {Boolean}
  */
-function isOpenTaskReport(report) {
-    return isTaskReport(report) && report.stateNum === CONST.REPORT.STATE_NUM.OPEN && report.statusNum === CONST.REPORT.STATUS.OPEN;
+function isCanceledTaskReport(report = {}, parentReportAction = {}) {
```
[the remark this answers is not in the document]
Comment (thienlnam) [reply src/libs/ReportUtils.js:206]: Yeah you are correct, a couple different report types can have the isDeletedParentAction. Though I am planning on keeping this name as task reports are slightly different in that the report can have the isDeletedParentAction key as well

**Comment by thienlnam** [comment]
Thanks for the quick review!

> There's no line below the parent action/deleted task

Fixed this!

> We are still showing the message of deleted task and the task title

Yeah this is expected for now, there's a different issue that will handle updating the system messages

**Comment by mollfpr** [comment]
@thienlnam The task detail is not immediately removed after I navigate to the task detail report with the browser back button from a thread. The issue is easier to observe offline. Here's the step:

1. Go offline
2. Create a task and open the task detail
3. Send a message
4. Right-click on the message and select reply in a thread
5. Press the browser back button
6. Cancel the task

<details>
<summary>OFFLINE</summary>

https://github.com/Expensify/App/assets/25520267/a9c8f56e-a73d-46e0-9a76-3330331483e6

</details>

<details>
<summary>ONLINE</summary>

https://github.com/Expensify/App/assets/25520267/de147dd3-c1d8-4021-a8f2-9d770a24e802

</summary>

**Comment by thienlnam** [comment]
> The task detail is not immediately removed after I navigate to the task detail report with the browser back button from a thread. The issue is easier to observe offline.

**Review by mollfpr — APPROVED** [review]
g75csv9111x41

**Review by tylerkaraszewski — APPROVED** [review]

**Comment by ayazhussain79** [comment]
Can we start testing canceled task flow?

**Comment by thienlnam** [comment]
@ayazhussain79 Yup! Feel free to start, we already created an issue for one known bug already here. But is ready to test now

**Comment by ayazhussain79** [comment]
Ok Thank you

**Comment by nam-odele** [comment]
@thienlnam after task deleted without any message on thread (except task action message) so in the parent report we still show [deleted message] or not.

**Comment by thienlnam** [comment]
>@thienlnam after task deleted without any message on thread (except task action message) so in the parent report we still show [deleted message] or not.

This is how threads work so we should align with that - this may require some BE changes though

File: src/languages/es.ts [hunk 3]
```diff
@@ -1542,7 +1542,7 @@ export default {
         completed: 'Completada',
         messages: {
             completed: 'tarea completada',
-            canceled: 'tarea cancelada',
+            canceled: 'tarea eliminado',
```
Comment (iwiznia) [inline src/languages/es.ts:1545]: @thienlnam  I don't know where you got this translation from (I checked slack and could not find it anywhere), but it is wrong. Please make sure you always get confirmation in #espanol-tranlsations

Comment (iwiznia) [reply src/languages/es.ts:1545]: Sending a PR to fix this

File: src/libs/actions/Task.js [hunk 4]
```diff
@@ -755,26 +755,50 @@ function getShareDestination(reportID, reports, personalDetails) {
...
+    const message = `deleted task: ${taskTitle}`;
     const optimisticCancelReportAction = ReportUtils.buildOptimisticTaskReportAction(taskReportID, CONST.REPORT.ACTIONS.TYPE.TASKCANCELLED, message);
     const optimisticReportActionID = optimisticCancelReportAction.reportActionID;
+    const taskReport = ReportUtils.getReport(taskReportID);
+    const parentReportAction = ReportActionsUtils.getParentReportAction(taskReport);
+    const parentReport = ReportUtils.getParentReport(taskReport);
+
+    const optimisticReportActions = {
+        [parentReportAction.reportActionID]: {
+            pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE,
```
Comment (alitoshmatov) [inline src/libs/actions/Task.js:767]: Pending action here could be either `update` or `delete` based on if it has thread messages

**Pull request merged**